src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.CompilationManager.cs

```diff
@@ -0,0 +1,218 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.ErrorReporting;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Diagnostics.EngineV2
+{
+    internal partial class DiagnosticIncrementalAnalyzer
+    {
+        /// <summary>
+        /// This cache CompilationWithAnalyzer for active/open files. 
+        /// This will aggressively let go cached compilationWithAnalyzers to not hold them into memory too long.
+        /// </summary>
+        private class CompilationManager
+        {
+            private readonly DiagnosticIncrementalAnalyzer _owner;
+            private ConditionalWeakTable<Project, CompilationWithAnalyzers> _map;
```

Comment:
By the way, does the Project instance change if additional file is edited? I know the compilation is unchanged, but we need to create a new CompilationWithAnalyzers instance as the AnalysisOptions have changed and hence needs reanalysis. Can you test what happens here if just additional file is changed? I don't remember if we have a test for it in DiagnosticServiceTests.vb